virt-handler: Restrict write access to nodes (#246)
kubevirt/community: design-proposals/shadow-node.md

## Goals

- Prevent a malicious user that has taken over a Kubernetes node where virt-handler daemonset is deployed from exploiting the daemonset's RBAC permissions in order to elevate privileges beyond the node until potentially having full privileged access to the whole cluster.
- Prevent virt-handler from modifying non-kubevirt-owned fields on the node. This includes not being able to change the spec or any label/annotation which are not strictly kubevirt owned.

## Non Goals

Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.   

## Definition of Users

Not user-facing change

## User Stories

Not user-facing change

## Repos

Kubevirt/Kubevirt

# Design

## High level design

The security issue is being addressed by introducing a new light weight CRD: ShadowNodes for each Kubernetes node.
Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.

Pros:

> - Since the new controller runs virt-controller context which only runs on the control-place nodes, the risk of taking over the cluster is greatly mitigated.

Reviewer: Changed.

   * This is triggered both by interval (3 min) and on changes of cluster config.

### Node writers on virt-controller
1. virt-controller/node-controller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-controller/watch/node.go#L318) the node unschedulable for VMs on heartbeat timeout.
2. virt-controller/nodetopologyupdater [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-controller/watch/topology/nodetopologyupdater.go#L63) the node with supported topology frequencies.

## Shadow node

The proposed solution is to introduce new CRD that will be shadowing a node "ShadowNode". Every entity that now writes to the node, will be writing to ShadowNode instead, and the new controller will have a responsibility to sync allowed information to Node.

## API Examples
```yaml
apiVersion: <apiversion>
kind: ShadowNode
metadata:
  name: <Equivalent to existing name>

>   # Labels and Annotations only

Reviewer 2: I think this is a great opportunity, to create a real spec for all that info. The reason we use for instance for heartbeat metadata is because we don't have a place on the k8s node object to properly express information. Now we can.

Reviewer: Added to the pros list to the CRD approach

Reviewer 3: Yes, having a typed spec with a well defined list of things that we want on the node, looks to be an opportunity here.

Reviewer: @rmohr how about if we make it a bit wider. Instead of setting a CRD for this very specific node sync purpose, define a virt-handler CRD, that would hold this `nodeInfo`  (such as the labels and annotations) and also future stuff that could be used in future virt-handler communication with other components...

Reviewer 2: I am fine either way.

  * Old virt-controllers: no upgrade issue
  * Old Virt-API: no upgrade issue
* New CRD, new RBAC; old virt-handler; old virt-controller; old virt-API:
  * Old virt-handler: during upgrade, the old RBACs are kept as backup until upgrade is done. Hence, patching node should not fail so there is no upgrade issue.
  * Old virt-controllers: no upgrade issue, virt controller RBAC is not changed due to this solution.
  * Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there is no upgrade issue.
  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed.
  * [same as before] Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node.
  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
  * [same as before] Old Virt-API: no upgrade issue.

### After upgrade rollout:

> Virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.

Reviewer 2: This reads like you would never stop trying to patch. In case I did not misread this, let's suppose we have 100 nodes, we update the node every 10s (can't remember the default interval), and we may be able to remove this code path in 12 months, that means we would now patch and fail for 12 months  with 10 calls/second. So that is definitely not an option. Just think about metrics collection for instance which will record the failed call on virt-handler as well as on the API server.

The only way would be to stop patching when you get the "forbidden" error, but then we have again the issue, that we can't differentiate a config error from a real desire to no longer patch it.

Reviewer: Indeed stop after first "forbidden" error is possible, but you are right that we can't differentiate a config error from a real desire to no longer patch it.

Cons:
- The status section is considered a user facing API section. Dynamically adding/updating 150+ technical labels there may only confuse the user and thus is not advisable.

### Using the virt-handler pod as a medium to pass the labels/annotations
Same proxy solution as primary solution where virt-handler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/annotations to the virt-handler pods's labels/annotations.
Labels/Annotations that need to pass to the node should get a special prefix (like "nodesync/"), so that regular labels don't get passed to the node by mistake.

Pros:
- Avoid introducing a new CRD to the project.
- A simple and clean approach.

Cons:
- It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.

### Creating a GRPC communication between virt-handler and virt-controller
Pass the labels/annotations through a grpc channel to the virt-controller.

Pros:
- Avoid introducing a new CRD to the project.

> - Allows for future expansions.

Reviewer 4: Do you have specific expansions in mind? This statement seems generic.

I would suggest to add here a security benefit: virt-handler does not need to access kube api at all, and cannot modify unrelated nodes or shadow nodes. virt-controller does all the risky work.

Reviewer: >Do you have specific expansions in mind? This statement seems generic.

it is. Will remove.

>I would suggest to add here a security benefit: virt-handler does not need to access kube api at all, and cannot modify unrelated nodes or shadow nodes. virt-controller does all the risky work.

Added.

### Using the virt-handler pod as a medium to pass the labels/annotations
Same proxy solution as primary solution where virt-handler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/annotations to the virt-handler pods's labels/annotations.
Labels/Annotations that need to pass to the node should get a special prefix (like "nodesync/"), so that regular labels don't get passed to the node by mistake.

Pros:
- Avoid introducing a new CRD to the project.
- A simple and clean approach.

Cons:
- It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.

### Creating a GRPC communication between virt-handler and virt-controller
Pass the labels/annotations through a grpc channel to the virt-controller.

Pros:
- Avoid introducing a new CRD to the project.
- security: Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work

Cons:

> - More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC case logic internally.

Reviewer 5: Perhaps I am missing something, but could you please explain what is more complex about it and why a CRD is more appealing?

virt-handler is already serving other gRPC commands for the virt-api, so I see no complexity to add another command to serve virt-controller/virt-operator.

The use of a CRD hints there is a need to expose the information, either for input or output for external projects to use/consume. But the requirement here is internal, nothing to do with an external project consuming this new CRD.

Reviewer: plainly, it's more complex and will take more time to implement than the suggested CRD option.

I personally don't see a problem with adding a CRD for internal use (no significant increase in storage, API calls, etc..).
However I agree that adding a new CRD for such a small purpose of proxying info to the node is an overkill, so it justifies looking for an alternative solution. Using GRPC is such a candidate solution.

Reviewer 5: Can you estimate the extra implementation effort in terms of time?

Reviewer: well, the effort would comprise of several steps (of course open for suggestion/improvements):
- on virt-handler:
  - replace every place we patch the node with a write to a new "cache", a place where all the new annotations/labels will be stored.
  - create or use an existing grpc/rest server, and use it to return this cache when asked. 
- on virt-controller:
  - add a completely new grpc client component, that will 
    - run only on the leader controller pod (similar to other controllers).
    - run as a go routine and poll the client for updates every Fr minutes. Fr will the lowest of the frequencies, meaning heartbeat (=1m IIRC). 
    - ask every virt-handler server separately (need to make sure it's possible to address them in a unique manner)
  - for each grpc response, thie client will
    - filter only the kubevirt-owned labels/annotations
    - patch them to the appropriate node accordingly.
   
Well, as to your question, I think it will take longer than the proposed solution. How much? maybe 1.5 as long? I would prefer to consult with other virt team members to be more confident.

This proposal is outlining mitigation in a form of internal change.

## Motivation

Kubevirt should be secure by default. While mitigation is available, it is not part of Kubevirt.

## Goals

- Prevent a malicious user that has taken over a Kubernetes node where virt-handler daemonset is deployed from exploiting the daemonset's RBAC permissions in order to elevate privileges beyond the node until potentially having full privileged access to the whole cluster.
- Prevent virt-handler from modifying non-kubevirt-owned fields on the node. This includes not being able to change the spec or any label/annotation which are not strictly kubevirt owned.

## Non Goals

Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.

## Definition of Users

Not user-facing change

## User Stories

Not user-facing change

## Repos

Kubevirt/Kubevirt

# Design Options

## Background
In order to easy review and understand the different options, this section will describe different usage of Node object in the virt-handler.

### Node writers on virt-handler
Kubevirt sets ~170 labels on each node. This is set by the following entities

> 1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node unschedulable for VMs at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs.

Reviewer 4: ```suggestion
1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node with `kubevirt.io/schedulable=false` at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs.
```

Reviewer: Fixed in order to be explicit.

Pros:
- Avoid introducing a new CRD to the project.
- Using existing web-server infra on the server side.
- security:
  - the web-service is secure using the cert-manager internal API.
  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
- Scalability:
  - Reduces #API requests (i.e. node patches) to a minimum.

Cons:
- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.
- Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.

## Solution option #3: Moving to third party entity
The idea is to eliminate the reason(s) to patch to the node from virt-handler in the first place.

> 1. Use tools like [Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/v0.15/get-started/introduction.html) or utilize virt-handler's [device-manager](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) to publish node features (pending checking whether they don't have the CVE issue as well).

Reviewer 4: I think it is worth to elaborate this idea a bit further. What do you intend to do with NFD and device plugins?

For example, setting `kubevirt.io/schedulable` may be replaced by tearing down the `devices.kubevirt.io/kvm` device.

Reviewer: What do you mean by tearing down?

If you mean that we can monitor the `devices.kubevirt.io/kvm` device from virt-controller and set the `kubevirt.io/schedulable` in case it is removed/unhealthy - then I don't think that is supported. Meaning a device can only be set unhealthy manually, and kubelet does not support removing devices if the plugin connection is lost

Reviewer 4: I'm asking to provide more detail on how you are going to use device manager and node feature discovery.

I raised an example, which you may be thinking about. I don't know what exposes the devices.kubevirt.io/kvm`, but that component may un-expose it if virt-handler is shut down, effectively making `kubevirt.io/schedulable` redundant.

Reviewer: The thing is that virt-handler **is** the one running device-manager and exposes these devices to the node, so it can't be used to monitor itself..
I honestly don't think it can be used to replace the heartbeat, but again I'm open for suggestions. 

I will elaborate more on how device-manager plugin can be used to replace the node-labeller.

Reviewer: DONE. PTAL

2. Remove writing the unschedulable label in case virt-handler crashes. It is racy, and has no real value.

> 3. Heartbeat - Remove adding the annotation in favor of watching the readiness of the virt-handler Deamonset/pods.

Reviewer 4: "remove adding" is mentally complex. Please explain what you intend to do instead of the heartbeat label. Are you going to abandon it?

Reviewer: in short- yes
I will rephrase

Reviewer 4: I wonder if you can elaborate. What is the suggested solution? Currently you light-heartedly suggest to "Use tools like NFD". To be a serious candidate, this alternative should propose actual specific tools, explain what they are going to do, how they can maintain needed functionality and upgrade considerations.

Reviewer: Will do

Reviewer: DONE. PTAL

Pros:
- No need to invent the wheel.
- Re-examine current methods to make sure if we still need them is a good practice. In this case we may find out that patching is no longer needed, and in the long run need to maintain less code.
- This steps can be broken to multiple seemingly unrelated PRs, that are easier to review.

Cons:

> - Relying on a third part tools.

Reviewer 4: I don't understand this statement: afaict KubeVirt already depends on Node Feature Discovery and Device Plugins.

Reviewer: virt-handler uses device-manger, but AFIK not NFD:
```
$ git grep node-feature-discovery
tests/vmi_lifecycle_test.go:                            By("adding a node-feature-discovery CPU model label to a node")
```

Reviewer 2: kubevirt is compatible with NFD, but all virt specific discovery, PCI device discovery and management, and much more is all done in kubevirt.

I don't think NFD is an option, and by moving it from an (initially external approach) into kubevirt we simplified the architecture, increase feature velocity and removed a lot of external failure points. Would advise against externalizing anything in that area, while of course not doing anything which would make kubevirt not being able to pick up NFD stuff in addition.

Reviewer: ACK. + NFD will not filter kubevirt-owned labels.

Reviewer 3: I agree with Roman, labels are too critical for KubeVirt in order to rely on an external project such as NFD.

Author: +1 in regards that moving this to NFD would be step back and on top of it this option seems like a lot of work (at least in comparison to other options)

Reviewer: ACK

    * On the other side - Virt-controller's node-ctonroller watches which monitors the heartbeat annotation.

3. Virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT...
  * This is triggered both by interval (3 min) and on changes of cluster config.

## Solution option #1: ShadowNode CRD
The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.
Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.

Pros:

> - Since the new controller runs virt-controller context which only runs on the control-plane nodes, the risk of taking over the cluster is reduced.

Reviewer 6: Are you sure this statement is correct?

On an openshift cluster, I see:
```bash
> k get pod -n openshift-cnv virt-controller-84d87bfdb6-8w55q -o yaml | grep nodename -i
  nodeName: cnvqe-14.lab.eng.tlv2.redhat.com
> k get pod -n openshift-cnv virt-controller-84d87bfdb6-wzkgf -o yaml | grep nodename -i
  nodeName: cnvqe-15.lab.eng.tlv2.redhat.com
> k get nodes
NAME                                      STATUS   ROLES                  AGE     VERSION
cnvqe-14.lab.eng.tlv2.redhat.com          Ready    worker                 7d22h   v1.28.7+f1b5f6c
cnvqe-15.lab.eng.tlv2.redhat.com          Ready    worker                 7d22h   v1.28.7+f1b5f6c
master-0.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
master-1.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
master-2.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
zeus21.lab.eng.tlv2.redhat.com            Ready    worker                 7d22h   v1.28.7+f1b5f6c

```

if indeed virt-controller is running on a worker node, I'm not sure I understand how the security risk is being reduced with this solution.

What am I missing?

Reviewer: If you are correct, then it is not. I need to recheck this.
Thanks for this notice!

Reviewer 2: You should not run kubevirt CP on worker nodes. That is a scurity issue on its own.

Reviewer 2: for example: if you take over virt-controller, you can create a pod in every namspace with an image of your choosing, mounting any secret you like and dumping it.

Reviewer 6: > You should not run kubevirt CP on worker nodes.

Interesting. I wasn't aware of it. Probably since I don't think I've even seen kubevirt running this way.

Quick questions:
1. Do you think we should default our infra components to run on master nodes?
2. Is this a common thing to do for operators? Are you aware of a k8s operator that runs pods on master nodes?

Reviewer: @iholder101 after checking, it looks like this is configurable on kubevirt CR via node-placement (see [docs](https://kubevirt.io/user-guide/operations/installation/#restricting-kubevirt-components-node-placement)) - however I don't think this configuration comes out of the box (relying only 2 samples of Itamar's cluster seen above and the SNO cluster I'm using..).

@rmohr - if this is a real concern - then 
- shouldn't node-placement configuration for infra components come out of the box? 
- in the end - can we, for the purpose of this proposal, rely on the assumption the virt-controller is indeed only deployed on control-plane nodes?

Reviewer 2: I think it would be a correct general improvement.

But independent how we install by default, if there is a separation between CP nodes and workload then kubevirt must follow that sepration or it is weakening security. For edge deployments they are typically not separated, but even there we could just follow the CP node roles and install fine.

Reviewer 2: Just one thing regarding to openshift, I think I remember vaguely that openshift had an infra node pool in addition to CP nodes where also no workloads are scheduled and that there was a plan to schedule them there (which may be ok too). Does my memory trick me there?

Reviewer: not that I know of, [docs](https://docs.openshift.com/container-platform/4.15/nodes/nodes/nodes-nodes-creating-infrastructure-nodes.html) don't seem to mention it either

Reviewer 3: The infra nodes concept is optional in OpenShift.

Reviewer 3: @iholder101 you grep only shows that `nodeName` is provided in the yaml.
But `nodeName` is a field in the spec, thus provided by somebody, not disclosed by the system.
IMU Kube is still not disclosing the actual host name, but only host ip. I'm not sure why, as security by obscurity is no security - but it's still the way it is.

For reference https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodStatus

Reviewer 2: > @iholder101 you grep only shows that `nodeName` is provided in the yaml.
> But `nodeName` is a field in the spec, thus provided by somebody, not disclosed by the system.
> IMU Kube is still not disclosing the actual host name, but only host ip. I'm not sure why, as security by obscurity is no security - but it's still the way it is.

I think there are two things to consider:

 1. I don't think that the node name has a relevancy here, only that our CP runs on non-CP nodes here is relevant. Getting access to worker nodes with compromised pods is much easier than getting access to CP nodes. Opening attack surfaces via angles like mentioned above if our CP runs there.
 2. While, again, node names don't play a role here, I can always get access to all nodes names where I can schedule by simply creating a daemonset and checking the pod status of all created pods. But maybe I don't get the point you try to make.

Reviewer 2: > The infra nodes concept is optional in OpenShift.

If so, I would highly recommend fixing the CP deployment for OpenShift Virtualization.

Reviewer 6: I forgot to update that this PR is ongoing to solve this issue: https://github.com/kubevirt/kubevirt/pull/11659.
It's still facing a few problems (regarding network lanes failures), but I'm actively working on it.

@fabiand 
> But nodeName is a field in the spec, thus provided by somebody, not disclosed by the system.

Unfortunately this is not correct. The `.spec.nodeName` fields acts as both a spec and a status field de-facto, and it will always be populated when the pod lands on a node.

- A malicious user can't manipulate virt-handler's RBAC permissions to exploit non-kubevirt params/labels.
- Adding a new CRD is a good opportunity to create a real spec for all that info. For instance, the reason we use heartbeat in metadata is that we don't have a place on the k8s node object to properly express information. Now we can.
- Scalability:
  - #shadowNodes will be equivalent #nodes, negligible space overhead

Cons:
- Scalability:

>   - Albeit a simple and lightweight CRD (will only contain metadata), adding shadowNodes adds a 1:1 shadowNodes to nodes resources, therefor increasing:

Reviewer 2: this is not a con. Going ti our own CR is an improvement over the current solution, especially because the heartbeat is not performed anymore on the node agent. We right now force very node object watcher to get every node object updated periodically based on our heartbeat interval. So this solution is an improvement for scalability. 

You mention that at the end somehow, but the "double write" is more than neglegtible compared to the watcher traffic which we cause today.

Reviewer: I agree it should be more highlighted. Added a distinct Pro for it.

Virt-Controller (Client): Add a new web-service client infra that will:
1. run as a go routine and poll the client for updates every one minute.This frequency is chosen because it is the highest of the frequencies (=heartbeat).
2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
3. filter non-kubevirt owned annotations/labels.
4. patch the node if necessary.

The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor for the web-service timeout mechanism (if received timeout - set node as unreachable).

Pros:
- Avoid introducing a new CRD to the project.
- Using existing web-server infra on the server side.
- security:
  - the web-service is secure using the cert-manager internal API.
  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
- Scalability:

>   - Reduces #API requests (i.e. node patches) to a minimum.

Reviewer 2: there is no difference to nod patch count with this approach.

Reviewer: there is if you take into account the heartbeat annotation removal (I am comparing with current state of course)

Reviewer 2: Ack.

Cons:
- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.

> - Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.

Reviewer 2: minimizing the window removed quite a bit of experienced issues around scheduling. Sacrificing this for an not preferrable approach does not sound like a good idea.

Reviewer: Hey @rmohr - I was not aware of this fixing user-experience issues. This would be a good reason to keep it. Can you specify which scenarios were involved?

I wonder how often does these scenarios occur (all these add the `schedulable: false` on teardown):
- virt-handler crashing
- heartbeat `stopCh` channel closing

The third scenario where virt-handler sets the label to false as soon as possible (until heartbeat sets it to "ready" later) - can maybe a good culprit. I would need to measure how much time it takes to be more certain.. 

But in any case I think that at the very least we can agree that setting the label to false on virt-handler is best-effort, and racy by design, so we need a good cause to keep it.

Reviewer 2: Maybe I misunderstood your pint. I thought you were considering changing that we set schedulable to false fast when shutting down (not crash). For crashes virt-controller already sets it to false.

Reviewer: my bad - I didn't distinguish between crash and shut-down.

What I mean is - instead of patching the node with the label like before, for crash scenarios we can for example utilize device-manager:
- when virt-handler crashes it severs the device-manager-to-kubelet grpc connection, so the devices change to 0, like so:
```
$ cluster-up/kubectl.sh get node node01 -ojson | jq .status.allocatable | grep kubevirt.io
selecting podman as container runtime
  "devices.kubevirt.io/kvm": "0",
  "devices.kubevirt.io/tun": "0",
  "devices.kubevirt.io/vhost-net": "0",
```
We can rely on this behavior and set the `schedulable: false` label according to `devices.kubevirt.io/kvm`'s quantity. This can be done by virt-controller's node-watcher as well..
- When virt-handler starts and registers to kubelet, it will initially set devices.kubevirt.io/kvm`'s quantity to 0 as well until virt-handler is ready (the first `schedulable: true` label was set at the first heartbeat cycle so that would be the time when it is set to the usual "1k" quantity) so that VMI won't get scheduled to that node before that.

Reviewer: Another way is to add a liveness probe to virt-handler, and when the virt-handler Demonset is not "ready" - then virt-controller will find out which virt-handler is missing and set that `schedulable: false` label

Pros:
- Avoid introducing a new CRD to the project.

Cons:
- [Deal-breaker] The status section is considered a user facing API section. Dynamically adding/updating 150+ technical labels there may only confuse the user and thus is not advisable.

## Solution option #5: [Ruled-out] Using the virt-handler pod as a medium to pass the labels/annotations
Same proxy solution as primary solution where virt-handler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/annotations to the virt-handler pods's labels/annotations.
Labels/Annotations that need to pass to the node should get a special prefix (like "nodesync/"), so that regular labels don't get passed to the node by mistake.

Pros:
- Avoid introducing a new CRD to the project.
- A simple and clean approach.

Cons:
- [Deal-breaker] It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.

> ## Solution option #6: [Ruled-out] Fine-tune each virt-handler RBAC to only patch its own node.

Reviewer: I agree. I mentioned this in the follow up steps section in the end of the proposal.

BTW I saw several interesting approaches to this that are worth looking into when it comes to the table: 
1. Using validatingadmissionpolicy https://github.com/openshift/origin/pull/28673
2. Using validating webhook as I think you also suggested IIRC (https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/node-identity.md)

> ## Solution option #6: [Ruled-out] Fine-tune each virt-handler RBAC to only patch its own node.

Reviewer 2: Yep 1 is a something to check out. 2 is something we have to avoid on the node object itself. As I mentioned it would put kubevirt in a critical cluster health path (but we can have either 1 or 2 on our shadow node cr, since we own the shadow node cr and we don't influence cluster health there with our webhook availability)

This proposal is outlining mitigation in a form of internal change.

## Motivation

Kubevirt should be secure by default. While mitigation is available, it is not part of Kubevirt.

## Goals

- Prevent a malicious user that has taken over a Kubernetes node where virt-handler daemonset is deployed from exploiting the daemonset's RBAC permissions in order to elevate privileges beyond the node until potentially having full privileged access to the whole cluster.
- Prevent virt-handler from modifying non-kubevirt-owned fields on the node. This includes not being able to change the spec or any label/annotation which are not strictly kubevirt owned.

## Non Goals

Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.

## Definition of Users

Not user-facing change

## User Stories

Not user-facing change

## Repos

Kubevirt/Kubevirt

# Design Options

## Background
In order to easy review and understand the different options, this section will describe different usage of Node object in the virt-handler.

### Node writers on virt-handler
Kubevirt sets ~170 labels on each node. This is set by the following entities

> 1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node with `kubevirt.io/schedulable=false` at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs.

Author: Maybe we should mentioned that this improves user experience with scheduling.

Reviewer: DONE

2. virt-handler/heartbeat:
  * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node with `kubevirt.io/schedulable=false` when the stop channel closes (i.e. virt-handler exits).
  * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
    * On the other side - Virt-controller's node-ctonroller watches which monitors the heartbeat annotation.

3. Virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT...

>   * This is triggered both by interval (3 min) and on changes of cluster config.

Author: This should be all static unless virt-handler is restarted, afaik. The reconciliation is more of best effort to prevent accidental removal of important labels.

Reviewer: added more info to make it clear.

## Solution option #1: Use Validating Admission Policy
Add a new [validationAdmissionPolicy and ValidatingAdmissionPolicyBinding](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/), adding the logic to:
1. prevent virt-handler from adding/deleting/updating non kubevirt-owned labels/annotations

> 2. (optional as it relies on alpha feature) prevent virt-handler from updating other nodes.

Author: What feature is alpha?

Reviewer: added.

Pros:
* Avoid adding CRDs or changing the current virt-handler code to fix the CVE issue.
* A k8s built-in feature which is straightforward to maintain and configure.
* validationAdmissionPolicy is [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it could be used for the next Kubevirt minor release.

Cons:
* The last validation (that prevents virt-handler from writing to other nodes) requires an alpha feature, so it can be excluded until it is GA.
  * Even without this validation, the CVE risk is greatly reduced.

## Solution option #2: ShadowNode CRD
The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.
Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.

Pros:

> - Since the new controller runs virt-controller context which only runs on the control-plane nodes, the risk of taking over the cluster is reduced.

Author: Let's rephrase " which only runs on the control-plane nodes,".
The virt-controller is running only on subset of nodes, therefore the exposure is not so wide as the virt-handler. Also for the controller we should highly recommend to run it on control-plane or a nodes that can be more isolated.

Reviewer: DONE.

- Scalability:
  - #shadowNodes will be equivalent #nodes, negligible space overhead
  - If heartbeat annotation is handled only on the shadowNode and not directly patched to the node as been done now, then the nodes will get much fewer API-calls. Ergo, Node-watching elements will get "bothered" less frequently.

Cons:
- Scalability:
  - Albeit a simple and lightweight CRD (will only contain metadata), adding shadowNodes adds a 1:1 shadowNodes to node resources, therefore increasing:
    - the #API-calls kubevirt does - At first #writes could double due to writing via the CRD as a proxy. However, this should be transiently reduced, as the labels kubevirt patches don't tend to change often (except the heartbeat that as mentioned above can be eliminated). Considering the heartbeat annotation removal from the node patch, then this con can be flagged as negligible.
    - the storage kubevirt takes.
- Adding virt-controller as a "proxy" agent, we now depend on its availability. This could cause an increase in heartbeat timeout in corner cases like upgrade.

> - To a degree, it is still possible to manipulate pods by causing VMs to migrate from other nodes.

Author: Please elaborate

Reviewer: DONE.

2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
3. filter non-kubevirt owned annotations/labels.
4. patch the node if necessary.

Notes:
- The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor using the web-service timeout mechanism (if received timeout - set node as unreachable).

Pros:
- Avoid introducing a new CRD to the project.
- Using existing web-server infra on the server side.
- security:
  - the web-service is secure using the cert-manager internal API.
  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
- Scalability:
  - Reduces #API requests (i.e. node patches) to a minimum.

Cons:
- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.

> - Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.

Author: Please correct the second sentence.

Reviewer: Added more info on this both on the Con and on the solution body.

Cons:
- [Deal-breaker] It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.

## Solution option #7: [Ruled-out] Fine-tune each virt-handler RBAC to only patch its own node.
Introducing a new controller called token-distributor-controller. It will create a token for each virt-handler pod and deliver it to the node.

Pros:
- Avoid introducing a new CRD to the project.
- A malicious user can't manipulate pods from other nodes to migrate to their node.

Cons:
- A malicious user can still manipulate its own node with non-kubevirt controlled params/labels, then possibly lure control plane components to its node in order to steal their credentials. For example, it can add the control-plane label to its own node, then add other labels to attract components that seek specific functionalities, and take over them.

This solution is not a bad one, but it's more important to address not being able to affect non-kubevirt labels/annotations before doing this (see Follow-up Steps section).

# chosen solution

TBD until further discussion.

> # Update/Rollback Compatibility

Author: TBD per chosen solution

Reviewer: Should I delete the stuff currently written?

2. virt-handler/heartbeat:
  * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node with `kubevirt.io/schedulable=false` when the stop channel closes (i.e. virt-handler exits).
  * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
    * On the other side - Virt-controller's node-controller watches which monitors the heartbeat annotation.

3. Virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT...
  * This is triggered both by interval (3 min) and on changes of cluster config. However, as these labels usually don't change—this intervalled reconciliation is usually no-op.

> ## Solution option #1: Use Validating Admission Policy

Reviewer 3: I like this approach, as to me it is taking us into the direction we should explore for us, but also for Kube: To limit node-local workloads to it's node.

This pattern here could help to evolve VAP in upstream, and serve as a pattern for other projects.

Reviewer 5: +1
This seems the most suitable solution out of the list presented here.

It will be great to see a PoC for it.

Reviewer: After I bump kubevirt/kubvirt k8s vendors to v0.30.0 (WIP) then I could set up such a PoC.
Until then please consider [this](https://playcel.undistro.io/?content=H4sIAAAAAAAAA%2B1da3PjNrL9K4w%2BbDJVJkcv27Kq%2FMGvycxev9ZyHO%2But1QQCUqM%2BBqAlO2k8t9vA6Qs6kFKaCpbt25hqnZjQjwHQANoNkg2zx8NhySk0W9Eo9%2BonfRfQgP%2Bkdh7oox7Udg3Zq2sbOqFTt%2B4jRyaHQc0IQKbQwAUhlFCEgDxjzLDsElMbC95t5IojvxZYHnR52bToS5J%2FaRvvDQ6vXbvsNU77vZaR82XRjVyNm5VY7hnzSI%2FDag1TUeUhTShXABDaLcH7f%2Fxj5fGgu%2Bl0X9p2OHM%2FE5NL3QZMdsdC46%2F07blk5FFw7HFnLRtMepMSGLZUfDS%2BPPHRX3THreiWWhFbPx5EvHEtD2Hcajm3y%2BNVtNqWa3DI6vd%2BdzuvjT%2BU4LzO%2BaYJPSVvJt2FLreOGtlbiJoIhwEwuyisXxCoC0vjYOXhhcmlLnEhqY78rcRM%2BnbULU7wBQQ2ySOwyjnkoh2%2B91OvzvqU7t%2FfNwXJhb1xfOTqDhtUw%2BXzpJUq6fAGSF9S8xJFK%2BRHOYU8xNWCA5zOJjCjCOWmDQkIz8zS8JSKn%2Bd%2BSSc2wOa%2FWfZYMHEeI3YFE7lq%2BbOgGU4Ubs9IZx73BQTqtMdHbbcbs9sdZots2v3Oubo%2BIgKI5BDu0Pb9lGngmo%2B8ixKYTRNn8Wm5woDZq3y4lk3N0PTOupaTav9uXW0pXmiWS%2BNdnFdrJ0TjIMks2Jh8PtGr91vt%2FtOu3886p%2B0%2Bj2ngiJmXkDYe1l7l8e9ssE8HcF4rA1EPjnaPeh283O7A7OjkiZhJOReYvJXL7EnWe9KjdnrbTUmo0GUUPP3KAR7eWMGQwWWVVxfG5klZUgCWoMO3NvMY4lwbhNKWDKiRPrTdrPdNZuHZrP12Gr1m8f99sm%2FlqZCATjlgTkhoePD1AtISMZUTh2X%2BJwWMTBHJl5IM%2FeUMunkrSimIZ94rmSCnxIW%2Bfew%2FOij8K%2FR%2BL1vDLxw7NMHGvueTZToUsZomFzk%2FhD%2BdihYARrJxTKx3eMOIXTUbp50em7z8OTk8MjuUjKizRNK6aFKVQ7lHlD%2FF6u6ZMSDC2sa2hFzotD8K6v0gfIsBvv%2F96u9G3HKZtQZiP9nF2dnH%2FEBco4J0sF7aMvBErMNpm02bA%2BURymz6UfQ8tLotjrdTqd3rFZFLK%2Fh8neT2PO2qnEwSjgGx8E44BAuwTcsYKHrWMK5rYQzLiz2lFET3AT1wXHacWrC%2FzzHOrt8PigcXQ1uvxWPn56Xj9rLh4et9vmva0UXl2tFl%2F9YK%2FqyVvJ0vV50u9Sei5v754uvP%2FcKRV%2Bufxl8HV63inV%2BuTnrFA%2BfBw8rh3f3j4WSb2ed9vDs4eLr8OLsvlh%2Bfn9eOLw%2B%2B1ps9M3l8OL66qxIfXP3dH5VPL4v2u9u8Dw4eyr%2BPri%2FuhhePD5cDweD82IPBo9Qd%2FH4n4OLs%2Bvr5ZKrItXTTbGq595x8ejnq8fzp1axZKUh8vhitWDZSrJokBXIyWdlEZUjiyDQdF4h2BwGcDX2kglMawfcufxNhKO%2B5ZLA898LBZ5TOJiBe4nYMC9jTmI9XD5e3DwWD687F2fF45vzlcObpcO724OpWAR%2BvkSt27vh139tKht%2B%2BQWMm%2F%2FAqe%2BF6dtH7%2FLiWeYsLDf1%2FdWygPwWsbVCL1wvZHAxFX8cBBAtQDQUpgE5mEeXnHnRzBIbGah4tTT3BNCg2PbMZqfZHLaa7ZEVQygG174DnkQMrshmGIUsd5zEdzw%2BPeDv4LADM%2BJDRn3wNtT6drmh8O7%2B6nbw9duXx%2BHT1cPgG1hvwzmDx4erq4oTHr5eXVf8nP8y3NiAxY%2B5QatPEebd6vmyi5U1m3v6FC5iLjhWZytS2H4LUo5MPkwbnfM3CJUWpycpoJfPG0VRYgcOTDg4m0%2Fp6zDx7Olpy0i4fQqd9sRMMJidxo5YcPDHEHodEH9IXOjXUOBPW4sqoDfR5PfTKDSyU%2B0RP%2B2anZODbtc8PslbADNk6PHIh7XCT5vin9tsHjSzf66RWd2x4GeoBboMm%2BrT5kHroH3QOeg2D7qtg277oNsp1io2l%2F7Qi4IgFZVnf8SJ8VFNHi4OPfb9oNAg0dqhWFGFVhZ5l8wQRjxyEz%2Byp2lshIE3fCUQuDvR%2BLRpBDY9jVx3yYZggoRRGNfMGw1hBxKdtpasle8ehpN0TGNoIP%2F9tPVz3ptYDuIpLCG5b%2FxAZTcM%2BMp0sT9ijHlobJIkgau46VDxH3GBz7adcyYbGiVW6aMHbAkJ4kU43jFbXRGOtzr9ZrMQjsNgEN%2F7Habkxy0T0yjc7wg%2FbrcYRn6tX7QbdjbU%2BZ88nL97hZlwLU6xChH%2BeiMNYwTNX%2BkrYaI%2FJHCOupXnRRBqSF9auOMC%2FjmPRhaLbr6%2F6Dhh9Ar%2BzKXJJoNtQZNRoI5x3tQxlKtjAsfkfOTggIk3itWRsWerg1i2K1QEwSyXVy3P9xIPY54ZYhhmb20MCKLK0SsSaCMGUAKd70igi8TNfCwwDD1l6CjwWhiQ%2BgDavuunXN055LgoVp%2Fetv86UgcF0Uwd9NZDYFpHyiCHKkMoC9TXtds6UvdB7si0Idpj6kAxwqbfUl%2BkbkDUMXGqjuFjPoI4Vh34xtXtMXY47AaUYZP3mDLYpUTqNcJ1Sn2peCOmPrMEyCS%2Bupfzwhnsn9SnCOAgGFWG%2BWTiDn31wAQBCWz1SQxhszrGQa7PADcfg0A9OADvO0L0LEbUhFiYQcLUQaF622KiboMYEQCiFlRs%2B0HqIwKj2LHVF0fsxLQ1VndO8Rhhwqn6lSFGRF1xFNshYrS4PSHhmJqBjVmOMXdiHA5hSU476tENc%2BzExzSROYyE6jOZOZxSDAwuKOr7OthFMthL4gaBU0SFUy8WUZXrm7OAhgl7V6cICKLeANPYmNqmnTD1wICrByGoHT1HrAPAqG%2BVANS11LdlEoapDdaqOgp1a4O%2FcxsT%2B2ECOMCYDiX5LVp18JA4v6Vc3Usn%2FA03j9PAU7foLFDv3AwRlwFGPiidecm7OfHVzbJE8ErE6xterH7plDSxZ89MRseeeERQg2KGiH4W6Lc26k5dgQCNt1nH9CPiwJWEvnm4wRAc4tETrUfSyxpSiyFrBprCobDU0Wh5YTQ90mlT%2BUSzBom0xCh0bHesfk1aoaEucmoXODBbglUKylxz7Ecj4uO82oIvjCSjQ0fpGOI8HBPiNmCOMzG7iDm2HeCx9E28DeKrB185AcE5KUDG%2BStyXjjGUcCaAqc9NT0YQDyDvOlQb10ImprL4oMCvSoWDHtZFIIujDiZ0ZprQhBJmnrGkRS1jJMxMEqDWDwfNRMvQDbI9embeOrLIFpAEUCIgr8keCE3ozTBzVWxVupUPcP6uAwq7q2KV2QS%2BlaLhstXOusy%2BfG4li3EDapaBLMYGePl2H2YU%2FLk9szeUsbwRObISzDb8RyNtmIQhR4EaHUIZk4dOGd1uh4k6o9iJU5uUtCtDgMPDxZvtMAmyZ%2BaXSQ%2B5TUC6jjAXYnEW7XUwUcLe7l0MCcOamwHsluJdfDinmIdfFJnN5Pdm8TjuYhVHZHuAeEI7gLIJ8SJXs1ZYCPxcjfoI57sCrQA4NsubB%2B5Lqc46wm4uMOFDbjTkFGeMM8Wy2icUsTtJ0EzkzFITEMH25BZ4KYh7t7ELHiFqI2KW84iJnQ96ju4wZiBD0UD63UfGzW8juB6X2P1v4kQWt1ayFtRb2OajGbqd5plK3EoRCMFDPP6T7Ut5cvrebqXeL97DX7OYKv3Sn3xQOpx8LwvHvPb%2BcMATXZBwL3AOWRKzSzfZo9UNTv6lfA9mGuJpZ6xvs3ez5nnlD8I3p2hXkNu6YT4tPTZ9674eo24ixPKonD4c%2Bl630pxT0P2HqLhAwis6o5IgaOeOQbTdzn1L3yPhsk%2BptucsbiY6jH%2BCgEAhMF4a80Jtjejnv%2Fbg9fbk4Oq45ZqOyOsC6rleHDupoaTQbsWlEPBu5GazmNfLmNfjgLpHpBOQe6%2FNz9Udxn9ntLQljdm282Tw%2BbxcbPZVGAQWySRmlNWeZZSuGFViPSeElCWu7MxHFqLEAvJP3EhnxaPFdm3ePRTjZqfnje%2B8bErWOYB1yS4uKxJcPmPmgRfauKfrusS3NYY%2F3luNJrgI5Maz3BztvEVoN3Az4OHWuC7%2B0c0fikDHM9yfn%2BOBovscjR4nouOJxCZ63j0Pd735FnxaPxyDj2eRmTc49FZfn4d%2FBXeBE83ePM%2F947x2Oy7Anh8rYHPvllQD17HZ2TfQ8DAl76egCAo%2BdYCgqn4ZQYBP1LGZnc4e4fKwI8PPvQNGZDtDF98GwLR4cKXJPDom%2FNa4Jsa4LtbRfCGr13UZ5Dfy8DRrHxdA0dS%2FBZH3zi0Wl2raXbbxxb1T6y33tGwmJi%2BO5386oRo0a6zedMXPwS%2B1cURzL8OIjiaO1IUPiSiaMyNnx2px%2FHxkRJFmo2fNFHkKP0AiirPhs%2Bl9A02sSOOxK99WUW0qGu1dvW2277D0je6rSPrpGuJr0eAGXutVrNjNpHcxU%2B4iIaeWLvO5qrvtdTu8urXYSThXtg%2BVu1S4%2BT3ObNbo%2BCAsc9QCizifof4ntv8URNf56z6TASCqSy5QpFql887qFFWpXgrMlXlfKtRladLq%2FFU50%2BrcZUlVKuybM6wVmcpS7lWZCrNpVbjqcpEVmXaz9iX5aYqsmxOIFUk2ZrmqcpXmnKoRlSVnqnGtC0VUpFtpzxHRc49rLs9ufHqXDRVrvLkNDWmsmw1NZaSVDRlkq25aTUYK5PVELxbs9ewnCXpbFi68peK0Iz7I9whAw5Hui0lDsO6JUcOR1mdNKfOWZlFp063S1odlnVbnl0t3rIkmlqkJWk19Th3yELCVrBDrh6CevNLfCiismw%2BHFlJeh%2BOrCrfD8W4OQEQRbUlIxDBuS1FEEm5NWcQybvvJb4tq7AW5V%2BzwHfMO0QyVyYi1uHcr4F3TlVU59%2BWu6jOWJXMqM5Wld2IYatId8TQleU%2FYrm2ZfChebdnSKKoq1ImUYSVOZQoxrL0CDTZXzJGu6RdIoir8jBRdPsbmm2ZmhjGqtRNBF9lLieCb3NyJ4KoMttTna8q%2FVOdbUs%2BKIawKkEUwbc5YxRBVJ1CiiD8S67x1UmmGL7KrFMMYWUaKoawKi8VyVeeqKpOuD1zFcFZncqKIKzKbVWnEwx77G519iuOryIdVp1wt%2FxYdd6tCbMIytIMWgzXjim1COqSHFsc055NuLcwc0tarjpheW6pGlcFj3wYvY4o%2Bwp52fnzbI2yx%2F8lsLLHHSWnw0U29L3xJKFhUPICUDm0zN%2BUAdJQXMlVIPw9LHnwUAXYmFlRiSiPMEpQqh1J%2FFGpjEIZBpZR6NCNz9mWX7bZqA6zfIqY3fX1LCuFZopNF7qVhfcN17QEl%2BcR8efmLDuH2xPqpKW5R8J0Jov81beQcokgMxaql6v8ZaBMwGrXszPRqk1nL%2BtORVy%2BeLvyUlveL%2FC8Gx8fL1K2zEW%2B1iYDJLmip7UiTYQdbvxUEZ8D7ht%2F%2FJkfJeBAF5JIxPcjmyRyGBeNj6Psotg%2BlG9gvoSR79xptWutdq3VrrXatVa71mrXWu1aq11rtWutdq3VrrXatVa71mrXWu1aq11rtev%2Fa2rXYn8y03rXqxitd70M0nrXVUCtd70ZqvWu10CB1rvWetdrGK13vQmk9a7XcVrv%2BgOm9a5ldK%2F1rnOg1rvOMVrveiNK612Xg7XedSlS610vYFrvuviel9a73ozRetda71rrXe%2F%2BYY4dSbTedRWf1rvGEWi9a613vZVI610vE2i9a613rfWui2itd62I03rXyjitd631rrXetda71nrXWu96M0rrXWu9a613jcNrvWutd40j0HrXWu96R5TWu9Z617HWu1bDar3rOgRa71rrXWPxWu9a611rvWtlrNa71nrXWu9aFaz1rrXetda7XqLRetda73pnKq13rfWuP5i03rXWu94Hk9a71nrXWu96Zy6td631rrXedTmp1rveD53Wu9Z616vUWu9a611rvevdKbXetda73pbMqPWutd51NaHWu9Z613hGrXeN4tN610girXe9J0Ktd70HTq13XZdP613vg0vrXddi0nrXWu9a612vQ5ZfttF615HWu%2F5%2FpHctO80ztWuWa0fm549ZlIotqDyYrYpgf5wsByCDS6oHPIscLaw9BI7HRHB9zAYTYOI3KGC53uYv95dnj1eiEHZ37FvoRnlDxWFWd%2FZCXV9odno2JbYdgTftz%2BfLxx9zvdgcLqZcs9V1R657QprHdlZM3xJWEAZPkwm4fs%2FOcpGXp7ec9rIFi%2FEylQ3xEjYOGuLmKuXC1I1%2B4%2FNn41fqw48U5q4BDTAurq6Ne5%2B8i7EJnR9eQjhluczwuEFCqRbG5EN%2BChyjM85pMPLfYRq9Cr1Ug4ZgiEhezgQ3VOtD7A5IR%2FxNmSd%2FefWSSVZvFARRaFx9tM64JuE4hY278RPU%2F8kSLZGtMY1fRTRnvMPskE1b9MiYQNX5Sb9wKokJODYDfhJ%2Fc8%2BhhhsxaHycJoYQZj%2BAv41%2Fnt1cG1D898Hdrfg9IEnOci%2BYjR8f0vBH2YsZ8VNSUjkZEy%2FkiaxpUUFOdJX3X%2BBsIZgmdVONyAUGEsQ%2B5XnDeOxl81GMF0BD%2BnobJWewZF9z4TJu9E%2BNSCrRW3N9eSvTPbNcz4dh%2BWl6YPwwFW8lJ6JJP700Ch7tpfFJEke%2Bs5F4rnKP5pbsjAoO6YeM7H0sWMF5o2XhabEqUbIClC5rAf3b3%2BZdzn9ZxsuyZYYcKsZ33hODp%2FbEIDB%2FHRFmF4TlDJrYlpVVs26ARc%2Bnxg%2Bn4CMzi7w0DDg%2FLyLhXLh3uTyXxfsiNx1Lv7BlPV6wnQWeWdSy3oR%2FT%2F8je7xW%2BmlTp0Wfwyg05yOTqeJBaCK8lewkh17%2FVGsOffokGpTx1Jwyn1Q6MXo3pvRdrFaxFmlxZiBrn1t9Kmqs2xejtDnzISyrQAwmGKLx5%2F8Cu0lKcVz9AAA%3D) CEL example.

Reviewer: update: a POC was performed.

# Overview

A reported [Advisory](https://github.com/kubevirt/kubevirt/security/advisories/GHSA-cp96-jpmq-xrr2) is outlining how of our virt-handler component can be abused to escalate local privileges when a node, running virt-handler, is compromised. A flow and more details can be found on reported [issue](https://github.com/kubevirt/kubevirt/issues/9109).

This proposal is outlining mitigation in a form of internal change.

## Motivation

Kubevirt should be secure by default. While mitigation is available, it is not part of Kubevirt.

## Goals

- Prevent a malicious user that has taken over a Kubernetes node where virt-handler daemonset is deployed from exploiting the daemonset's RBAC permissions in order to elevate privileges beyond the node until potentially having full privileged access to the whole cluster.
- Prevent virt-handler from modifying non-kubevirt-owned fields on the node. This includes not being able to change the spec or any label/annotation which are not strictly kubevirt owned.

## Non Goals

> Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.

Reviewer 5: IMO, this **is** the proper solution for the long-run. Being the goal Kubevirt should strive to in order to be secure.

Several of the existing solutions that require patching the node should be reevaluated and optimized not to require access to nodes (or other cluster level core component).

Said that, I understand this may take more resources and time to reach, therefore the solutions provided in this proposal are still required. I just hope to see a continuation to this effort.

Reviewer: I agree that some of the suggestions should occur anyway in parallel, as part of the goal of keeping kubevirt working properly.
However, these suggestions deserve a separate proposal, and for this specific CVE they are not strictly required.

Reviewer 6: I agree with both of you :) it should be explored but a bit out-of-scope.

One thing we should consider is whether the CPU information on the nodes is essential (I think the vast majority of our labels are dedicated to CPU info). After all, this information is fixed by nature and stays the same for each node, maybe except for very rare cases. Currently we use these labels as selectors, but we can save this information elsewhere (e.g. a configmap?), then virt-controller would target nodes by name (or the hostname label?) before creating the launcher pod.

But as you said it's out-of-scope for this PR, just something to keep in mind.

The validationAdmissionPolicy and ValidatingAdmissionPolicyBinding objects will be configured with the following parameters:
- [matchConstraints](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#getting-started-with-validating-admission-policy): apply only on Nodes; Operations: UPDATE
- predefined [variables](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#variable-composition):
  - `isPartitionedServiceAccount`: true if virt-handler is the user requesting the update.
  - `oldLabels`: the nodes' labels before the update was issued
  - `oldNonKubevirtLabels`: a sublist of the old labels that are not kubevirt owned
  - `newLabels`: the nodes' labels after the update was issued
  - `newNonKubevirtLabels`: a sublist of the new labels that are not kubevirt owned
  - (same set of parameters for annotations)
- [matchConditions](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#matching-requests-matchconditions) rules (only if all are "true" then the validation will be evaluated):
  - `isPartitionedServiceAccount` == true

> - [validation](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#creating-a-validatingadmissionpolicy) rules (written in pseudocode to keep things readable):

Reviewer 5: Is it possible to express something like: "reject-all-changes except these"?
I mean, without the need to compare other labels/annotation we do not care about.

Reviewer: well, it was nice if we only got the "changes" we made, but it doesn't work like that AFAIK: When setting a POST request (this is true for both webhooks and validationAdmissionPolicies) you can add validation rules to validate [these](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#validation-expression) entities (such as `object`, `oldObject`, `request`  - see more details [here](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response) to learn how they look like)
so it's logical that you first have to gather the kubevirt-labels on both {`object`,`oldObject`} (or the non-kubevirt-own, as shown) , and then make sure that we "reject-all-changes except these"..
Please review this CEL example that models the validations that the valiationAdmissionPolicy will have [here](https://playcel.undistro.io/?content=H4sIAAAAAAAAA%2B1da3PjNrL9K4w%2BbDJVJkcv27Kq%2FMGvycxev9ZyHO%2But1QQCUqM%2BBqAlO2k8t9vA6Qs6kFKaCpbt25hqnZjQjwHQANoNkg2zx8NhySk0W9Eo9%2BonfRfQgP%2Bkdh7oox7Udg3Zq2sbOqFTt%2B4jRyaHQc0IQKbQwAUhlFCEgDxjzLDsElMbC95t5IojvxZYHnR52bToS5J%2FaRvvDQ6vXbvsNU77vZaR82XRjVyNm5VY7hnzSI%2FDag1TUeUhTShXABDaLcH7f%2Fxj5fGgu%2Bl0X9p2OHM%2FE5NL3QZMdsdC46%2F07blk5FFw7HFnLRtMepMSGLZUfDS%2BPPHRX3THreiWWhFbPx5EvHEtD2Hcajm3y%2BNVtNqWa3DI6vd%2BdzuvjT%2BU4LzO%2BaYJPSVvJt2FLreOGtlbiJoIhwEwuyisXxCoC0vjYOXhhcmlLnEhqY78rcRM%2BnbULU7wBQQ2ySOwyjnkoh2%2B91OvzvqU7t%2FfNwXJhb1xfOTqDhtUw%2BXzpJUq6fAGSF9S8xJFK%2BRHOYU8xNWCA5zOJjCjCOWmDQkIz8zS8JSKn%2Bd%2BSSc2wOa%2FWfZYMHEeI3YFE7lq%2BbOgGU4Ubs9IZx73BQTqtMdHbbcbs9sdZots2v3Oubo%2BIgKI5BDu0Pb9lGngmo%2B8ixKYTRNn8Wm5woDZq3y4lk3N0PTOupaTav9uXW0pXmiWS%2BNdnFdrJ0TjIMks2Jh8PtGr91vt%2FtOu3886p%2B0%2Bj2ngiJmXkDYe1l7l8e9ssE8HcF4rA1EPjnaPeh283O7A7OjkiZhJOReYvJXL7EnWe9KjdnrbTUmo0GUUPP3KAR7eWMGQwWWVVxfG5klZUgCWoMO3NvMY4lwbhNKWDKiRPrTdrPdNZuHZrP12Gr1m8f99sm%2FlqZCATjlgTkhoePD1AtISMZUTh2X%2BJwWMTBHJl5IM%2FeUMunkrSimIZ94rmSCnxIW%2Bfew%2FOij8K%2FR%2BL1vDLxw7NMHGvueTZToUsZomFzk%2FhD%2BdihYARrJxTKx3eMOIXTUbp50em7z8OTk8MjuUjKizRNK6aFKVQ7lHlD%2FF6u6ZMSDC2sa2hFzotD8K6v0gfIsBvv%2F96u9G3HKZtQZiP9nF2dnH%2FEBco4J0sF7aMvBErMNpm02bA%2BURymz6UfQ8tLotjrdTqd3rFZFLK%2Fh8neT2PO2qnEwSjgGx8E44BAuwTcsYKHrWMK5rYQzLiz2lFET3AT1wXHacWrC%2FzzHOrt8PigcXQ1uvxWPn56Xj9rLh4et9vmva0UXl2tFl%2F9YK%2FqyVvJ0vV50u9Sei5v754uvP%2FcKRV%2Bufxl8HV63inV%2BuTnrFA%2BfBw8rh3f3j4WSb2ed9vDs4eLr8OLsvlh%2Bfn9eOLw%2B%2B1ps9M3l8OL66qxIfXP3dH5VPL4v2u9u8Dw4eyr%2BPri%2FuhhePD5cDweD82IPBo9Qd%2FH4n4OLs%2Bvr5ZKrItXTTbGq595x8ejnq8fzp1axZKUh8vhitWDZSrJokBXIyWdlEZUjiyDQdF4h2BwGcDX2kglMawfcufxNhKO%2B5ZLA898LBZ5TOJiBe4nYMC9jTmI9XD5e3DwWD687F2fF45vzlcObpcO724OpWAR%2BvkSt27vh139tKht%2B%2BQWMm%2F%2FAqe%2BF6dtH7%2FLiWeYsLDf1%2FdWygPwWsbVCL1wvZHAxFX8cBBAtQDQUpgE5mEeXnHnRzBIbGah4tTT3BNCg2PbMZqfZHLaa7ZEVQygG174DnkQMrshmGIUsd5zEdzw%2BPeDv4LADM%2BJDRn3wNtT6drmh8O7%2B6nbw9duXx%2BHT1cPgG1hvwzmDx4erq4oTHr5eXVf8nP8y3NiAxY%2B5QatPEebd6vmyi5U1m3v6FC5iLjhWZytS2H4LUo5MPkwbnfM3CJUWpycpoJfPG0VRYgcOTDg4m0%2Fp6zDx7Olpy0i4fQqd9sRMMJidxo5YcPDHEHodEH9IXOjXUOBPW4sqoDfR5PfTKDSyU%2B0RP%2B2anZODbtc8PslbADNk6PHIh7XCT5vin9tsHjSzf66RWd2x4GeoBboMm%2BrT5kHroH3QOeg2D7qtg277oNsp1io2l%2F7Qi4IgFZVnf8SJ8VFNHi4OPfb9oNAg0dqhWFGFVhZ5l8wQRjxyEz%2Byp2lshIE3fCUQuDvR%2BLRpBDY9jVx3yYZggoRRGNfMGw1hBxKdtpasle8ehpN0TGNoIP%2F9tPVz3ptYDuIpLCG5b%2FxAZTcM%2BMp0sT9ijHlobJIkgau46VDxH3GBz7adcyYbGiVW6aMHbAkJ4kU43jFbXRGOtzr9ZrMQjsNgEN%2F7Habkxy0T0yjc7wg%2FbrcYRn6tX7QbdjbU%2BZ88nL97hZlwLU6xChH%2BeiMNYwTNX%2BkrYaI%2FJHCOupXnRRBqSF9auOMC%2FjmPRhaLbr6%2F6Dhh9Ar%2BzKXJJoNtQZNRoI5x3tQxlKtjAsfkfOTggIk3itWRsWerg1i2K1QEwSyXVy3P9xIPY54ZYhhmb20MCKLK0SsSaCMGUAKd70igi8TNfCwwDD1l6CjwWhiQ%2BgDavuunXN055LgoVp%2Fetv86UgcF0Uwd9NZDYFpHyiCHKkMoC9TXtds6UvdB7si0Idpj6kAxwqbfUl%2BkbkDUMXGqjuFjPoI4Vh34xtXtMXY47AaUYZP3mDLYpUTqNcJ1Sn2peCOmPrMEyCS%2Bupfzwhnsn9SnCOAgGFWG%2BWTiDn31wAQBCWz1SQxhszrGQa7PADcfg0A9OADvO0L0LEbUhFiYQcLUQaF622KiboMYEQCiFlRs%2B0HqIwKj2LHVF0fsxLQ1VndO8Rhhwqn6lSFGRF1xFNshYrS4PSHhmJqBjVmOMXdiHA5hSU476tENc%2BzExzSROYyE6jOZOZxSDAwuKOr7OthFMthL4gaBU0SFUy8WUZXrm7OAhgl7V6cICKLeANPYmNqmnTD1wICrByGoHT1HrAPAqG%2BVANS11LdlEoapDdaqOgp1a4O%2FcxsT%2B2ECOMCYDiX5LVp18JA4v6Vc3Usn%2FA03j9PAU7foLFDv3AwRlwFGPiidecm7OfHVzbJE8ErE6xterH7plDSxZ89MRseeeERQg2KGiH4W6Lc26k5dgQCNt1nH9CPiwJWEvnm4wRAc4tETrUfSyxpSiyFrBprCobDU0Wh5YTQ90mlT%2BUSzBom0xCh0bHesfk1aoaEucmoXODBbglUKylxz7Ecj4uO82oIvjCSjQ0fpGOI8HBPiNmCOMzG7iDm2HeCx9E28DeKrB185AcE5KUDG%2BStyXjjGUcCaAqc9NT0YQDyDvOlQb10ImprL4oMCvSoWDHtZFIIujDiZ0ZprQhBJmnrGkRS1jJMxMEqDWDwfNRMvQDbI9embeOrLIFpAEUCIgr8keCE3ozTBzVWxVupUPcP6uAwq7q2KV2QS%2BlaLhstXOusy%2BfG4li3EDapaBLMYGePl2H2YU%2FLk9szeUsbwRObISzDb8RyNtmIQhR4EaHUIZk4dOGd1uh4k6o9iJU5uUtCtDgMPDxZvtMAmyZ%2BaXSQ%2B5TUC6jjAXYnEW7XUwUcLe7l0MCcOamwHsluJdfDinmIdfFJnN5Pdm8TjuYhVHZHuAeEI7gLIJ8SJXs1ZYCPxcjfoI57sCrQA4NsubB%2B5Lqc46wm4uMOFDbjTkFGeMM8Wy2icUsTtJ0EzkzFITEMH25BZ4KYh7t7ELHiFqI2KW84iJnQ96ju4wZiBD0UD63UfGzW8juB6X2P1v4kQWt1ayFtRb2OajGbqd5plK3EoRCMFDPP6T7Ut5cvrebqXeL97DX7OYKv3Sn3xQOpx8LwvHvPb%2BcMATXZBwL3AOWRKzSzfZo9UNTv6lfA9mGuJpZ6xvs3ez5nnlD8I3p2hXkNu6YT4tPTZ9674eo24ixPKonD4c%2Bl630pxT0P2HqLhAwis6o5IgaOeOQbTdzn1L3yPhsk%2BptucsbiY6jH%2BCgEAhMF4a80Jtjejnv%2Fbg9fbk4Oq45ZqOyOsC6rleHDupoaTQbsWlEPBu5GazmNfLmNfjgLpHpBOQe6%2FNz9Udxn9ntLQljdm282Tw%2BbxcbPZVGAQWySRmlNWeZZSuGFViPSeElCWu7MxHFqLEAvJP3EhnxaPFdm3ePRTjZqfnje%2B8bErWOYB1yS4uKxJcPmPmgRfauKfrusS3NYY%2F3luNJrgI5Maz3BztvEVoN3Az4OHWuC7%2B0c0fikDHM9yfn%2BOBovscjR4nouOJxCZ63j0Pd735FnxaPxyDj2eRmTc49FZfn4d%2FBXeBE83ePM%2F947x2Oy7Anh8rYHPvllQD17HZ2TfQ8DAl76egCAo%2BdYCgqn4ZQYBP1LGZnc4e4fKwI8PPvQNGZDtDF98GwLR4cKXJPDom%2FNa4Jsa4LtbRfCGr13UZ5Dfy8DRrHxdA0dS%2FBZH3zi0Wl2raXbbxxb1T6y33tGwmJi%2BO5386oRo0a6zedMXPwS%2B1cURzL8OIjiaO1IUPiSiaMyNnx2px%2FHxkRJFmo2fNFHkKP0AiirPhs%2Bl9A02sSOOxK99WUW0qGu1dvW2277D0je6rSPrpGuJr0eAGXutVrNjNpHcxU%2B4iIaeWLvO5qrvtdTu8urXYSThXtg%2BVu1S4%2BT3ObNbo%2BCAsc9QCizifof4ntv8URNf56z6TASCqSy5QpFql887qFFWpXgrMlXlfKtRladLq%2FFU50%2BrcZUlVKuybM6wVmcpS7lWZCrNpVbjqcpEVmXaz9iX5aYqsmxOIFUk2ZrmqcpXmnKoRlSVnqnGtC0VUpFtpzxHRc49rLs9ufHqXDRVrvLkNDWmsmw1NZaSVDRlkq25aTUYK5PVELxbs9ewnCXpbFi68peK0Iz7I9whAw5Hui0lDsO6JUcOR1mdNKfOWZlFp063S1odlnVbnl0t3rIkmlqkJWk19Th3yELCVrBDrh6CevNLfCiismw%2BHFlJeh%2BOrCrfD8W4OQEQRbUlIxDBuS1FEEm5NWcQybvvJb4tq7AW5V%2BzwHfMO0QyVyYi1uHcr4F3TlVU59%2BWu6jOWJXMqM5Wld2IYatId8TQleU%2FYrm2ZfChebdnSKKoq1ImUYSVOZQoxrL0CDTZXzJGu6RdIoir8jBRdPsbmm2ZmhjGqtRNBF9lLieCb3NyJ4KoMttTna8q%2FVOdbUs%2BKIawKkEUwbc5YxRBVJ1CiiD8S67x1UmmGL7KrFMMYWUaKoawKi8VyVeeqKpOuD1zFcFZncqKIKzKbVWnEwx77G519iuOryIdVp1wt%2FxYdd6tCbMIytIMWgzXjim1COqSHFsc055NuLcwc0tarjpheW6pGlcFj3wYvY4o%2Bwp52fnzbI2yx%2F8lsLLHHSWnw0U29L3xJKFhUPICUDm0zN%2BUAdJQXMlVIPw9LHnwUAXYmFlRiSiPMEpQqh1J%2FFGpjEIZBpZR6NCNz9mWX7bZqA6zfIqY3fX1LCuFZopNF7qVhfcN17QEl%2BcR8efmLDuH2xPqpKW5R8J0Jov81beQcokgMxaql6v8ZaBMwGrXszPRqk1nL%2BtORVy%2BeLvyUlveL%2FC8Gx8fL1K2zEW%2B1iYDJLmip7UiTYQdbvxUEZ8D7ht%2F%2FJkfJeBAF5JIxPcjmyRyGBeNj6Psotg%2BlG9gvoSR79xptWutdq3VrrXatVa71mrXWu1aq11rtWutdq3VrrXatVa71mrXWu1aq11rtev%2Fa2rXYn8y03rXqxitd70M0nrXVUCtd70ZqvWu10CB1rvWetdrGK13vQmk9a7XcVrv%2BgOm9a5ldK%2F1rnOg1rvOMVrveiNK612Xg7XedSlS610vYFrvuviel9a73ozRetda71rrXe%2F%2BYY4dSbTedRWf1rvGEWi9a613vZVI610vE2i9a613rfWui2itd62I03rXyjitd631rrXetda71nrXWu96M0rrXWu9a613jcNrvWutd40j0HrXWu96R5TWu9Z617HWu1bDar3rOgRa71rrXWPxWu9a611rvWtlrNa71nrXWu9aFaz1rrXetda7XqLRetda73pnKq13rfWuP5i03rXWu94Hk9a71nrXWu96Zy6td631rrXedTmp1rveD53Wu9Z616vUWu9a611rvevdKbXetda73pbMqPWutd51NaHWu9Z613hGrXeN4tN610girXe9J0Ktd70HTq13XZdP613vg0vrXddi0nrXWu9a612vQ5ZfttF615HWu%2F5%2FpHctO80ztWuWa0fm549ZlIotqDyYrYpgf5wsByCDS6oHPIscLaw9BI7HRHB9zAYTYOI3KGC53uYv95dnj1eiEHZ37FvoRnlDxWFWd%2FZCXV9odno2JbYdgTftz%2BfLxx9zvdgcLqZcs9V1R657QprHdlZM3xJWEAZPkwm4fs%2FOcpGXp7ec9rIFi%2FEylQ3xEjYOGuLmKuXC1I1%2B4%2FNn41fqw48U5q4BDTAurq6Ne5%2B8i7EJnR9eQjhluczwuEFCqRbG5EN%2BChyjM85pMPLfYRq9Cr1Ug4ZgiEhezgQ3VOtD7A5IR%2FxNmSd%2FefWSSVZvFARRaFx9tM64JuE4hY278RPU%2F8kSLZGtMY1fRTRnvMPskE1b9MiYQNX5Sb9wKokJODYDfhJ%2Fc8%2BhhhsxaHycJoYQZj%2BAv41%2Fnt1cG1D898Hdrfg9IEnOci%2BYjR8f0vBH2YsZ8VNSUjkZEy%2FkiaxpUUFOdJX3X%2BBsIZgmdVONyAUGEsQ%2B5XnDeOxl81GMF0BD%2BnobJWewZF9z4TJu9E%2BNSCrRW3N9eSvTPbNcz4dh%2BWl6YPwwFW8lJ6JJP700Ch7tpfFJEke%2Bs5F4rnKP5pbsjAoO6YeM7H0sWMF5o2XhabEqUbIClC5rAf3b3%2BZdzn9ZxsuyZYYcKsZ33hODp%2FbEIDB%2FHRFmF4TlDJrYlpVVs26ARc%2Bnxg%2Bn4CMzi7w0DDg%2FLyLhXLh3uTyXxfsiNx1Lv7BlPV6wnQWeWdSy3oR%2FT%2F8je7xW%2BmlTp0Wfwyg05yOTqeJBaCK8lewkh17%2FVGsOffokGpTx1Jwyn1Q6MXo3pvRdrFaxFmlxZiBrn1t9Kmqs2xejtDnzISyrQAwmGKLx5%2F8Cu0lKcVz9AAA%3D)
Please play with it and if you find a better way to do it I'm all ears of course :)

  | size(`newNonKubevirtLabels`) == size(`oldNonKubevirtLabels`)                                                                      | virt-handler user cannot add/delete non kubevirt-owned labels                  |                                                                                                                                                                        |
  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |

### Backwards compatibility:
validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.

> However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by a new feature gate "node restriction". HCO main and v1.3 branches will enable this feature gate by default.

Reviewer 5: Can we program the component that deploys Kubevirt (virt-operator?) to condition what and how to setup the cluster based on its version/capability?
Clusters that can support the new policy (e.g. in v1.30 or in v1.29 where the K8S FG is enabled), will have Kubevirt deployed in one way and in other clusters in the existing manner.
The condition will be dropped once we drop support for K8S 1.29.

Having a FG protecting this change (or any other) will drag this for at least 2 version (6 months), requiring users to act on making their cluster more secure. I hope we can take that burden on ourself by just giving them the best possible option given the setup.

If this is not possible, could you please elaborate on why not?

Reviewer: I'm not sure it's possible, will need to check. 
But theoretically speaking, isn't that problematic to introduce such a dependency to a k8s version? It sounds unusual..

Reviewer: After thinking it through, I agree with you.
a FG is not a good fit for this, since this is not something we want a user (=cluster-admin) to need to switch on, as this is not a user facing feature. 
We have 2 alternatives then:
- add a logic on runtime to check k8s version and fail only if the vesion is above 0.30.0
- always deploy and ignore on fail.

I prefer the former. Will fix the proposal accordingly

### Next step: Prevent patches from other nodes' virt-handlers
It is possible to expand the rules of the validationAdmissionPolicy to also prohibit virt-handler running on a node from patching other nodes.
This could be defined by setting a new variable `requestHasNodeName` which will be true if request has `userInfo.extra.authentication.kubernetes.io/node-name` field, then adding this validation:
```
!`requestHasNodeName` OR (object.Name == request.nodeName)
```

> However, this field is [beta](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#additional-metadata-in-pod-bound-tokens) on k8s 1.30  and is currently protected by the [ServiceAccountTokenPodNodeInfo](https://github.com/kubernetes/kubernetes/blob/c4bce63d9886e5f1fc00f8c3b5a13ea0d2bdf772/pkg/features/kube_features.go#L753) k8s feature gate.

Reviewer 5: I guess this is considered a cons. Can you add it as such?

Reviewer: Added.

## Non Goals

Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.

## Definition of Users

Not user-facing change

## User Stories

Not user-facing change

## Repos

Kubevirt/Kubevirt

# Design Options

## Background
In order to easy review and understand the different options, this section will describe different usage of Node object in the virt-handler.

### Node writers on virt-handler
Kubevirt sets ~170 labels on each node. This is set by the following entities
1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node with `kubevirt.io/schedulable=false` at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs. This is meant to improve user experience with VM scheduling.

2. virt-handler/heartbeat:
  * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node with `kubevirt.io/schedulable=false` when the stop channel closes (i.e. virt-handler exits).

>   * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.

Reviewer: Thanks! DONE

  |-----------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
  | object.spec == oldObject.spec                                                                                                     | virt-handler user cannot modify spec of the nodes                              |                                                                                                                                                                        |
  | object.status == oldObject.status                                                                                                 | virt-handler user cannot modify status of the nodes                            |                                                                                                                                                                        |

>   | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |

Reviewer: IIUC, managed fields is a legitimate way to configure the object, therefore it shouldn't be restricted..

>   | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |

Reviewer: hmmm I'm not so sure anymore it is needed. @xpivarc - can you re-explain why this is needed?

  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |

### Backwards compatibility:
validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.
However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by the k8s version deployed. 
The virt-operator deployment logic should be: 
```
if k8s-ver >= v1.30 
then
  deploy the objects
else
  deploy, but only log on failure
```

> Once k8s v1.29 and below is not supported on kubevirt, this logic should be removed.

Reviewer 6: What if Kubevirt is being installed on k8s 1.29 (where validationAdmissionPolicy is beta), but the feature gate is enabled? I think it's reasonable to deploy the objects on such scenario, and it might even be an acceptable workaround for Kubevirt users until they're able to upgrade thier k8s version.

Perhaps we can take a different approach of using dry-run to determine if the object is known or not.
That is:
* Create the objects with a dry-run option. Check if an error is returned.
  * If a nil error is returned - re-create without the dry-run option.
  * If an "object unknown" error is returned ([this one](https://github.com/kubernetes/kube-openapi/blob/f0e62f92d13f418e2732b21c952fd17cab771c75/pkg/util/proto/validation/errors.go#L78)?), log and continue.
  * If a different error is returned - fail.

We can obviously drop the second step once we drop support for k8s 1.29.

Reviewer: Hey, I'm not sure why we need to dry run - if the object is not installed it'll not deploy anyway right? . 

But we do want to distinguish two behaviors:
For 1.29-, we want to deploy (or at least try), if we get error, log it.
If it's 1.30+, we do want to fail if errors, and not only log.

Reviewer 6: > Ignoring an "object not found" error in the general case is wrong imo

I agree, but relying on k8s version is (as someone also [said here](https://github.com/kubevirt/community/pull/246#discussion_r1586582058) :wink:) also "unusual" :) The question is what's worse.

But TBH I initially thought that for 1.29- we're going to not even to deploy the objects, which sounded problematic. If you're going to simply try to create the objects on a best-effort basis and log on error it sounds better to me.

Thanks!

### Next step: Prevent patches from other nodes' virt-handlers
It is possible to expand the rules of the validationAdmissionPolicy to also prohibit virt-handler running on a node from patching other nodes.
This could be defined by setting a new variable `requestHasNodeName` which will be true if request has `userInfo.extra.authentication.kubernetes.io/node-name` field, then adding this validation:
```
!`requestHasNodeName` OR (object.Name == request.nodeName)
```

> However, this field is [beta](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#additional-metadata-in-pod-bound-tokens) on k8s 1.30  and is currently protected by the [ServiceAccountTokenPodNodeInfo](https://github.com/kubernetes/kubernetes/blob/c4bce63d9886e5f1fc00f8c3b5a13ea0d2bdf772/pkg/features/kube_features.go#L753) k8s feature gate.

Reviewer 6: As I suggest in one of the other comments, WDYT about supporting it on a best-effort basis? Admins that run older k8s versions might want to consider using the FG as a workaround.

Reviewer: Wrote at that comment, I don't think we should best effort on versions we know it should work on, but it's open for debate,and besides that's a minor implementation detail, we can continue this important discussion on the future PR.

This field can be added once it is GA.

Pros:
* Avoid adding CRDs or changing the current virt-handler code to fix the CVE issue.
* A k8s built-in feature which is straightforward to maintain and configure.
* validationAdmissionPolicy is [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it could be used for the next Kubevirt minor release.

Cons:
* Currently, can't add the next step that prevents patches from other nodes' virt-handlers.

## Solution option #2: ShadowNode CRD
The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.

> Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.

Reviewer 7: Actually, I'm considering whether this solution could be viable, particularly if it's used alongside the [scheduling readiness feature of k8s](https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/), which is expected to be [GAed in k8s 1.30](https://github.com/kubernetes/enhancements/issues/3521#issuecomment-1843817972). What are your thoughts on this approach?

The flow I've been thinking about goes like this:

1. virt-handler is deployed and sets the `ShadowNode` Objects with the proper annotations and labels, just like you mentioned.
2. Whenever a vmi is created, virt-api will mutate the virt-launcher pod with a dedicated SchedulingGate.
3. virt-controller will detect virt-launcher pods with the SchedulingGate and do the following for each launcher pod:
    * Calculate suitable nodes for scheduling from the kubevirt perspective, let's mark it as \<nodeList\>.
    * Remove the SchedulingGate from the launcher pod and add node affinity to allow scheduling only to nodes in \<nodeList\>.

By leveraging scheduling readiness, we might be able to avoid node modifications entirely within KubeVirt, as suggested in [this comment](https://github.com/kubevirt/community/pull/246/files#r1586066156), and will also allow us to remove all the boilerplate and implementation details of KubeVirt from the nodes.  Additionally, it might address concerns about users with node access gaining visibility into hardware information. 
Any thoughts?

FYI @vladikr

Reviewer 8: @Barakmor1 I think this is a very interesting idea! I'm not 100% sure we need a gate. 
Virt-controller is the one that creates the pods, so it can calculate the list of suitable nodes before posting the Pod - and do the same during migration.
One possible issue is (as @xpivarc pointed out offline) that this list may become extensive. 

@rmohr FYI

Reviewer 7: > @Barakmor1 I think this is a very interesting idea! I'm not 100% sure we need a gate.
Virt-controller is the one that creates the pods, so it can calculate the list of suitable nodes before posting the Pod - and do the same during migration.

 but if no node match a launcher it will remain pending forever as the <nodeList> will remain empty.

> One possible issue is (as @xpivarc pointed out offline) that this list may become extensive.

I think that [the maximum amount of nodes is 5000](https://kubernetes.io/docs/setup/best-practices/cluster-large/)

Reviewer: My 2 cents - +1 to all efforts to use alternative ways to avoid the need to patch the node. 
If we could find good alternatives to all the places virt-handler is patching the node - then we could remove the patch permission entirely.
But even if we remove the node-labeller component we  currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.

Reviewer 7: > But even if we remove the node-labeller component we currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.

Can't we patch the `shadowNode` with heartbeat labels as well?

Reviewer: > > But even if we remove the node-labeller component we currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.
> 
> Can't we patch the `shadowNode` with heartbeat labels as well?

We can, but I believe we can solve it in a simpler, more k8s approach. Will issue a proposal about it in the near future.

# Overview

A reported [Advisory](https://github.com/kubevirt/kubevirt/security/advisories/GHSA-cp96-jpmq-xrr2) is outlining how of our virt-handler component can be abused to escalate local privileges when a node, running virt-handler, is compromised. A flow and more details can be found on reported [issue](https://github.com/kubevirt/kubevirt/issues/9109).

This proposal is outlining mitigation in a form of internal change.

## Motivation

Kubevirt should be secure by default. While mitigation is available, it is not part of Kubevirt.

## Goals

> Mitigate the advisory by default.

Reviewer 9: Can we please expand this to add explicit goals? I think @rmohr summarized the goals clearly in the comment.

Reviewer 3: +1 please expand

Reviewer: DONE

## Non Goals

## Definition of Users

Not user-facing change

## User Stories

Not user-facing change

## Repos

Kubevirt/Kubevirt

> # Design

Reviewer 9: One alternative design that was proposed the sig-scale call, that may solve the same problem is as follows:

1. Introduce a new controller, lets call it token-distributor-controller.
2. The token-distributor-controller has a RBAC Template CR that creates a unique ClusterRole, ClusterRolebinding and ServiceAccount for each virt-handler pod. Example of the template from kubevirt-handler clusterrole

```
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - list
  - watch
  - get
- apiGroups:
  - ""
  resources:
  - nodes
  resourceNames:
  - <node-name>
  verbs:
  - patch
```
3. The token-distributor-controller will then create a token for the pod via `TokenRequest` API [1][2] pod on the node and deliver the token on the node. 
4. The virt-handler pod will use a local volume mount to receive the token. An init container could be used to make sure the file is delivered by token-distributor-controller
5. Virt-handler will start with restricted permissions that could be carefully modified so that it will work on all the objects that are related to the node

This could be a generic solution that could help all daemonsets that are deployed with privileged RBAC and could be a separate project all together.  

Href:
1. https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount 
2. https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/

Reviewer 2: I don't think that this is addressing the probelm fully. There are two thing which need to be fixed:

 1. being able to modify other nodes (that can be addressed like above)
 2. Not being able to modify non-kubevirt-owned fields on the node (this includes not being able to change the spec or any label/annotation which is not stricty kubevirt owned)
 
 
 We could definitely create unique roles for each node, and in addition deploy a node webhook which would ensure that a service account only, that is also what the apiserver does internally for each kubelet, but that has the severe disadvantage that we would now need a **required** webhook on a core k8s resource. If our node webhook is ever unavailable, no kubelet, or anythig else can actually continue to modify the node. That's a no-go.

THere are other positive side-effects if we fix it with the "shadow nodes":

 * heartbeat can be moved to the shadow node, meaning we create much less load on all the controllers watching nodes.
  * we finally have a clear api to express our node-needs and don't need to press our API in annotations and labels on nodes.
 
 
 One may now think that now virt-controller still can modify all nodes, and that is true. However now we only have one component which can do that instead of n, further a `token-distributor-controller` controller which would create those rbac roles has to have the exact same permissions as well, otherwise it would not be able to create the roles in the first place ...

Reviewer 9: @rmohr thanks for summarizing the reasons. 

@xpivarc if it is helpful, please add this to alternative designs considered for future us to refer back at.

Reviewer 3: Can we includes this in the gola section?

@RamLavi @xpivarc ?

Reviewer: [alaypatel07](https://github.com/alaypatel07) thanks for bringing this up. 
Added the alternative to the proposal body.
Added to the goals.

Reviewer 3: @rmohr we have to address a third element

3. Prevent that the daemonset can modify other KubeVirt owned CRs related to other nodes

IOW 3. Restrict the damemonset to only modify KubeVirt owned CRs related to it's own node

Reviewer 2: Yes that third element is true, but I would qualify it as a follow-up step, not needed to address the CVE. Now virt-handler cannot blindly modify the node topology anymore. It can only modify hardware capability availability. It can not lure in privileged or sensitive workloads anymore by just applying labels which they are waiting for, or by making a node a control-plane node.

The follow up would be either cert based auth for virt-handler, or maybe we can still use tokens and get the IP in the webhooks for verification. Something like this ...

Let's add it to the doc as follow-up.

Reviewer: @fabiand AFAIK the only kubevirt CR virt-handler can update is `virtualmachineinstances`. Is this what you are referring to?

@rmohr I will add a `Follow-up Steps` section in the end of the doc and include:
- (the one just suggested) Restrict the virt-handler daemonset to only modify KubeVirt owned CRs related to it's own node
- Restrict the virt-handler daemonse from being able to update/patch other nodes

Is that alright with you?

Reviewer 3: @RamLavi yes - if this is the only one, then yes. And: generally speaking: handler should just be able to touch anything which a) it should be permitted to touch and b) only if it's own it's "own node"

Reviewer: ACK. DONE.

> (This should be brief and concise. We want just enough to get the point across)

Reviewer 3: To me a high level design is missing here.

It could be something like: The security issue is being addressed by introducing ShadowNodes for every Kubernetes node. virt-handlers are only interacting with ShadowNodes - which allows us to drop any node access privieleges from the handler's SA - a controller will then map these ShadowNode requests to the real kubernetes nodes, concentrating the required permissions on the controller.

Reviewer: thanks! DONE.

    2. As part of [do](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L119) we do 2 calls:
     [Get](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) - irrelevant as this is read access.
    [Patch](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) - updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.

3. As part of node labeller
    1. [run](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L195) - irrelevant as this is read access

    2. [patchNode](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) - node labeller patching node with labels for various informations which are mostly static in nature:  cpu model, hyperv features, cpu timer, sev/-es, RT...

> This is triggered both by interval (3 min) and on changes of cluster config. 

Reviewer 9: I wonder if the consumers for each of the above node labels and annotations should be mentioned alongside?

For example, the heartbeat in point 2. is consumed by the NodeController in virt-controller-app to check if node is healthy and raise an event.

Reviewer: I think adding the consumers will not necessarily add clarity here. I did rephrase to make it more high level.

> ## Shadow node

Reviewer 3: @rmohr @xpivarc @RamLavi do we know how other project shave solved this problem?

Reviewer: @fabiand I was not present during the time discussions on how other projects solved the issue. For obvious reasons, this is not something a project will advertise, so perhaps @xpivarc and @rmohr could shed more light on this. However, I did come across what I think is cilium's [fix](https://github.com/cilium/cilium/pull/22163) on the matter, and it looks like they went by the CRD approach, same as us.

Reviewer 3: FWIW I'd expect any author of such a feature to do some field research. Thus, yes, maybe you joined later, but that should not stop you from doing your own research :)

> ## Shadow node

Reviewer 3: @xpivarc iirc then this idea was taken to kube sig api- has it? what was the outcome of their discussion?

## API Examples
```yaml
apiVersion: <apiversion>
kind: ShadowNode
metadata:
  name: <Equivalent to existing name>
  // Labels and Annotations only
spec: {} //Empty, allows futher extension
status: {} //Empty, allows futher extension
```

## Scalability

There are few aspects to consider. 

1.  #shadowNodes will be equivalent #nodes, negligible space overhead

2. #writes could double. Here it is importent to look at the usage and sort each case to 2 categories (low and high frequent write). The first usage is clearly low frequent as long as virt-handler operates as designed.
The second usage consist of two cases which might seem different but in fact these are same most of the time because NodeSchedulable, CPUManager, KSMEnabledLabel labels and KSMHandlerManagedAnnotation annotation is mostly static. What is left is VirtHandlerHeartbeat that is not necessary to sync on a Node (requires re-working node controller).
The third case is propagating labels that are also mostly static.
With all the above doubling #writes is unlikely. 

## Update/Rollback Compatibility

> Virt-handler is going to continue writting to Node object, therefore update should be without compatibility issues.

Reviewer 9: When upgrade rolls out, virt-controllers are upgraded first then the daemonsets followed by virt-api. There will be a timeperiod where new virt-controllers (with ShadowNode code) will work with old virt-handlers and old-apiserver without ShadowNode API. Can we mention here clear steps how:
1. new virt-controllers will work with older virt-handlers and olde virt-api
2. new virt-controllers and newer virt-handlers will work with older virt-api

Reviewer: Hey @alaypatel07, I'll be taking over addressing the issue this proposal is trying to mitigate. I plan to take the same course this proposal is choosing. I will answer your question in the next reply, but if you think I should re-admit the proposal with the answers addressed in them - please tell me and I will do so.

Reviewer: In regard to your question, I think the upgrade order is a bit different. Please keep me honest, but according to [this](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/apply/reconcile.go#L526-L641) - virt-operator upgrade order is like this:
```
CRDs --> RBACs --> Daemonsets (=virt-handler) --> Controller Deployments (=virt-controller)--> API Deployments (=virt-API)
```
Also note that as mentioned above, the virt-handler will still keep trying to patch the node alongside the shadowNode, and at some point during the upgrade, it would start being blocked due to the removal of the patch ability. In order to mitigate this, the node patch will start ignoring RBAC "forbidden" error.

Having said the above, this is a specific mention to each upgrade scenario - before and after the upgrade itself:
 
 During upgrade rollout:
* New CRD; old RBAC; old virt-handler; old virt-controller; old virt-API:
  * Old RBAC: no upgrade issue
  * Old virt-handler: no upgrade issue
  * Old virt-controllers: no upgrade issue
  * Old Virt-API: no upgrade issue
* New CRD, new RBAC; old virt-handler; old virt-controller; old virt-API:
  * Old virt-handler: during upgrade, the old RBACs are kept as backup until upgrade is done. Hence, patching node should not fail so there is no upgrade issue.
  * Old virt-controllers: no upgrade issue, virt controller RBAC is not changed due to this solution.
  * Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there are no upgrade issue.
  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed. 
[same as before] Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node. 
  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
  * [same as before] Old Virt-API: no upgrade issue.

After upgrade rollout:
virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.

Reviewer 2: My suggestion, if possible, would be the following:

 * Introduce a new status field in the KubeVirt CR
 * Let virt-handlers continue patching the node as well
 * After virt-handlers and virt-controller are rolled over, set in the kubevirt CR a status field indicating that shadow nodes are now the way to go
  * virt-handler would pick it up right away over our config logic
  * finally remove the RBAC permissions.

Reviewer 2: There may be some neglectible overlap between virt-handlers picking up the config change and the rbac removal then, but that should worst case just result in a few rbac permission errors which we should then _not_ mask.

Reviewer: Hey @rmohr, thank you for your suggestion. I understand from what you say that you are concerned with masking the error. Am I right?
The way I see it, our options are:
1. (your suggestion) Adding a new status field. virt-operator will set it to true once [infrastructureRolledOver](https://github.com/kubevirt/kubevirt/blob/4179837bd7c20d991ae7a65d021b1f9a5480e1ae/pkg/virt-operator/resource/apply/reconcile.go#L519) is true. Everytime virt-handler wants to patch the node, it will check first if that status is true. 

Pros: 
    - avoid ignoring an error.
Cons: 
    - adding a field that will stay forever just for our little short-lived cause.
    - this has a potential for pesky corners issues (What if the upgrade failed somehow after the status was set? etc..)
    - there is a time gap until the config gets updated to each virt-handler via the informer - as you mentioned.

2. Checking before patching the node. Using the [SelfSubjectAccessReview](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/self-subject-access-review-v1/#SelfSubjectAccessReview) API in order to check if it's allowed to patch the node before doing it.

Pros: 
    - avoids ignoring an error.
    - robust and stateless. We don't care when the upgrade rollout is done. Instead we check the API server directly. 
Cons: 
    - [minor] need to add RBAC to virt-handler ("authorization.k8s.io", "SelfSubjectAccessReview", "create", "get", "delete").
    - there is the small window between us checking "am I allowed to patch node" and actually doing the patch, where things could change and end up with the forbidden error anyway.

3. (The current proposal) Try patching, ignoring if failing. no need to say more. 

Pros: 
    - robust and stateless. We don't care when the upgrade rollout is done. Instead we try and maybe fail.
Cons: 
    - masking an error.
    - hard to predict if will cause issues in the future (forward compatibility)

4. (optimizing the current proposal) Upon failing once to patch on "forbidden" error - stop patching the node. This is what Lubo did on the high frequency hearbeat labelling in his original [PR](https://github.com/kubevirt/kubevirt/pull/10131/commits/ab963f50144b16784d093cd04ac1137d0832bff5#diff-dd05fa462e124061de87457f1df303157a338be8a163ea85e91aeffa6fa40721R170)

I personally tend towards the SelfSubjectAccessReview option, or masking the error like originally suggestion, since the error we are ignoring is pretty specific, and I don't see how it can conflict in the future.

Reviewer 2: Hm, feels like SelfSubjectAccessReview way would just be another (expensive) way to just mask errors again (so we again just hope that it is correct that the permissions are gone). An extra status field would  from that perspective be much cheaper ...

Whenever we talk about masking we actually don't know if it is expected and I personally prefer to not add things where we just hope that things are right and swallow something.

Reviewer: @rmohr, I respectfully disagree. Using SelfSubjectAccessReview isn't just masking errors; it's a direct and intentional way of asking the RBAC server, "Can I patch the node?" rather than just hoping things are right. It cheaper than patching the actual node then failing, and ensures actions are only taken when permitted, adding a layer of security. 

I believe relying on the upgrade mechanism to signal a status field once ready could prove to be more costly in the long run, considering the potential corner scenarios mentioned earlier.

Reviewer 2: > @rmohr, I respectfully disagree. Using SelfSubjectAccessReview isn't just masking errors; it's a direct and intentional way of asking the RBAC server, "Can I patch the node?" rather than just hoping things are right.

I am not sure. I tried to express a knowledge gap which we try to overcome with checking for forbidden error or the SelfSubjectAccessReview. In both cases we don't  know if it is intentional and correct that we don't have the permission. With a clear config switch we know: Now I should do both, and from now on I should do one. Having a very brief period regarding to eventual consistency is fine where maybe one or two update requests would fail (like a low amount of conflict REST calls is fine too).

The information gap is what I consider "masking". Therefore the reviews add up quite a big front of additional uncertainties without making anything better (actually imo much more complex).

> It cheaper than patching the actual node then failing, and ensures actions are only taken when permitted, adding a layer of security.

I am not sure what "cheap" means here, and giving additional permissions which may leak info which we did not have access to before opens security questions which I don't even want to have to think about.

Reviewer: @alaypatel07  I updated the upgrade scenarios on the proposal's body as you suggested.

Reviewer 2: > @rmohr I am still against using the upgrade as a sign to know if we can stop patching the node. The upgrade process is complex and state-full, and relying on it may introduce corner cases. In a way, Lubo's "try and ignore if fails on specific error" is much more stateless and clean.
Can we agree to continue with his approach?

It should basically be a check like this:

```
if freshInstall && supportsShadowNodes()  {
  kubevirtCR.Status.UseShadowNode = True
}

if fully installed && !switchedToShadowNode && supportsShadowNodes() {
  kubevirtCR.Status.UseShadowNode = True
} else {
  kubevirtCR.Status.UseShadowNode = False
}
```

where `supportsShadowNodes()` can basically just check if the install strategy of the current version includes the shadow node CR, to ensure that the version we are upgrading downgrading to, has them.
